languagetool-core/src/main/java/org/languagetool/rules/spelling/hunspell/HunspellRule.java

```diff
@@ -181,59 +181,16 @@ protected boolean isQuotedCompound (AnalyzedSentence analyzedSentence, int idx,
             messages.getString("desc_spelling_short"));
           ruleMatch.setType(RuleMatch.Type.UnknownWord);
           if (userConfig == null || userConfig.getMaxSpellingSuggestions() == 0 || ruleMatches.size() <= userConfig.getMaxSpellingSuggestions()) {
-            List<SuggestedReplacement> suggestions = SuggestedReplacement.convert(getSuggestions(cleanWord));
-            if (word.endsWith(".")) {
-              int pos = 1;
-              for (String suggestion : getSuggestions(word)) {
-                if (!suggestions.contains(suggestion)) {
-                  suggestions.add(Math.min(pos, suggestions.size()), new SuggestedReplacement(suggestion.substring(0, suggestion.length()-1)));
-                  pos += 2;  // we mix the lists, as we don't know which one is the better one
+            ruleMatch.setLazySuggestedReplacements(() -> {
+              try {
+                synchronized (this) { // some getSuggestions overrides are thread-unsafe
+                  return calcSuggestions(word, cleanWord);
                 }
+              } catch (IOException e) {
+                logger.error("Error while calculating speller suggestions", e);
+                return Collections.singletonList(new SuggestedReplacement("Internal error: " + e.getMessage()));
```

Comment:
Do you know when this can happen? I wonder whether it makes sense to give out the error. in general, we tend to throw an exception instead.

Reply:
The exception can be thrown from `init()`, but it seems `init` should be already done by the time `getSuggestions` is called (at least in the current codebase, I can imagine someone calling `getSuggestions` explicitly without preceding `match`). So if you believe it's safe, I can remove `needsInit` checks from `getSuggestions`.

My idea with message was to give users indication that something went wrong. I can rethrow a `RuntimeException` if you think it's better. `MorfologikSpellerRule#appendLazySuggestions` would need to be changed as well.

Reply:
> I can rethrow a RuntimeException if you think it's better.

Yes, I think it's better to go with the "fail early" approach that serves us well so far.